### DISCUSSION TITLE
How to efficiently edding data with dependent Relationships in SQLAlchemy: Avoiding Overhead from Repeated Queries

### DISCUSSION THREAD

**Discussion by iancont** [sqlalchemy/sqlalchemy#11510] [Usage Questions]
Hi everyone,

I have a question about adding data to a normalized database using SQLAlchemy. Specifically, I'm wondering if I need to normalize the data first or if SQLAlchemy will handle the corresponding foreign keys automatically. How do I establish these add data relationships effectively?
This question applies when you want to add an instance of a class that depends on an attribute of another class, for which you may already have an existing instance.
Here's an example of what I'm working with:

```py
from sqlalchemy import create_engine, Column, Integer, String, ForeignKey
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, sessionmaker

Base = declarative_base()

class User(Base):
    __tablename__ = 'user'
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    addresses = relationship("Address", back_populates="user")

class Address(Base):
    __tablename__ = 'address'
    id = Column(Integer, primary_key=True)
    email = Column(String(100))
    user_id = Column(Integer, ForeignKey('user.id'))
    user = relationship("User", back_populates="addresses")

engine = create_engine('sqlite:///:memory:', echo=True)
Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session = Session()

def add_address_for_user(session, user_name, email):
    # Search for the user by name
    user = session.query(User).filter_by(name=user_name).first()

    # If user does not exist, create a new user
    if not user:
        user = User(name=user_name)
        session.add(user)
        session.commit()  # Commit to ensure the user is added and gets an ID

    # Create a new address for the user
    new_address = Address(email=email, user=user)
    session.add(new_address)
    session.commit()  # Commit to save the new address

# Adding the initial user
john_doe = User(name="John Doe")
session.add(john_doe)
session.commit()

# Now, add a new address for "John Doe"
add_address_for_user(session, "John Doe", "john.new@example.com")

# Verify the result
user = session.query(User).filter_by(name="John Doe").first()
print(f"User: {user.name}")
for address in user.addresses:
    print(f"Address: {address.email}")
```

I have been normalizing the data using Pydantic and performing a query for each field. However, this approach introduces significant overhead. Some instances of my models can take up to 5 seconds to create due to extensive querying of the database for distinct tables.

My main question is whether there is a simpler or more efficient way to handle this in SQLAlchemy. Do I need to create a custom method like `add_address_for_user`, or does SQLAlchemy provide a built-in or easier approach to managing these relationships and ensuring data normalization?

**Comment by zzzeek** [comment]
I dont know that I'm following, there's a lot of terminology used here that I can't match up with what you're doing.  Adding an Address - by "normalize", you mean match it to the correct user when you INSERT it?   You do need to associate the object with the correct parent explicitly, SQLAlchemy would have no way of knowing that.   I dont see what "query for each field" means, in your code I see one extra query for User which you could skip since you have that User already:

```py
def add_address_for_user(session, user, email):

    # Create a new address for the user
    new_address = Address(email=email, user=user)
    session.add(new_address)
    session.commit()  # Commit to save the new address

# Adding the initial user
john_doe = User(name="John Doe")
session.add(john_doe)
session.commit()

# Now, add a new address for "John Doe"
add_address_for_user(session, john_doe, "john.new@example.com")
```

im not sure what you mean otherwise
